amethyst_renderer/src/hide_system.rs

```diff
@@ -0,0 +1,130 @@
+use amethyst_core::{
+    specs::prelude::{
+        BitSet, InsertedFlag, ReadExpect, ReadStorage, ReaderId, RemovedFlag, Resources, System,
+        WriteStorage,
+    },
+    transform::components::{HierarchyEvent, Parent, ParentHierarchy},
+};
+use HiddenPropagate;
+
+// Based on the [UiTransformSystem](struct.UiTransformSystem.html).
+/// This system adds a [HiddenPropagate](struct.HiddenPropagate.html)-component to all children.
+/// Using this system will result in every child being hidden.
+/// Depends on the resource "ParentHierarchy", which is set up by the [TransformBundle](struct.TransformBundle.html)
+#[derive(Default)]
+pub struct HideHierarchySystem {
+    marked_as_modified: BitSet,
+
+    inserted_hidden_id: Option<ReaderId<InsertedFlag>>,
+    removed_hidden_id: Option<ReaderId<RemovedFlag>>,
+
+    parent_events_id: Option<ReaderId<HierarchyEvent>>,
+}
+
+impl<'a> System<'a> for HideHierarchySystem {
+    type SystemData = (
+        WriteStorage<'a, HiddenPropagate>,
+        ReadStorage<'a, Parent>,
+        ReadExpect<'a, ParentHierarchy>,
+    );
+    fn run(&mut self, (mut hidden, parents, hierarchy): Self::SystemData) {
+        #[cfg(feature = "profiler")]
+        profile_scope!("hide_hierarchy_system");
+
+        self.marked_as_modified.clear();
+
+        hidden.populate_inserted(
+            &mut self.inserted_hidden_id.as_mut().unwrap(),
+            &mut self.marked_as_modified,
+        );
+        hidden.populate_removed(
+            &mut self.removed_hidden_id.as_mut().unwrap(),
+            &mut self.marked_as_modified,
+        );
+
+        for event in hierarchy
+            .changed()
+            .read(&mut self.parent_events_id.as_mut().unwrap())
+        {
+            match *event {
+                HierarchyEvent::Removed(entity) => {
+                    self.marked_as_modified.add(entity.id());
+                }
+                HierarchyEvent::Modified(entity) => {
+                    // HierarchyEvent::Modified includes insertion of new components to the storage.
+                    self.marked_as_modified.add(entity.id());
+                }
+            }
+        }
+
+        // Compute hide status with parents.
+        for entity in hierarchy.all() {
+            {
+                let self_dirty = self.marked_as_modified.contains(entity.id());
+
+                let parent_entity = parents.get(*entity).unwrap().entity;
+                let parent_dirty = self.marked_as_modified.contains(parent_entity.id());
+                if parent_dirty {
+                    if hidden.contains(parent_entity) {
+                        for child in hierarchy.all_children_iter(parent_entity) {
+                            match hidden.insert(child, HiddenPropagate::default()) {
```

Comment:
blabla.map_err(|e| error!("msg {}", e))

Reply:
using `map_err` does not help, as it returns a `Result<T,E>` which has to be handled again.
The `?` operator is not implemented for `amethyst::error`

Reply:
Then you can
```rs
if let Err(e) = hierarchy.all_chi....{

}
```